include/ttmlir/Dialect/TTIR/Analysis/ShardSolver.h

```diff
@@ -0,0 +1,301 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef TTMLIR_DIALECT_TTIR_ANALYSIS_SHARDSOLVER_H
+#define TTMLIR_DIALECT_TTIR_ANALYSIS_SHARDSOLVER_H
+
+#include "ttmlir/Dialect/TT/IR/TTOpsTypes.h"
+#include "ttmlir/Dialect/TTIR/Analysis/Edge.h"
+#include <algorithm>
+#include <bitset>
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace mlir::tt::ttir {
+
+struct ShardSpec;
+
+struct ShardSolverSolution {
+  llvm::DenseMap<Operation *, LayoutAttr> selectedOpLayout;
+  std::unordered_set<Edge> reshardedEdges;
+
+  ShardSolverSolution(
+      const llvm::DenseMap<Operation *, LayoutAttr> &selectedOpLayout,
+      const std::unordered_set<Edge> &reshardedEdges)
+      : selectedOpLayout(selectedOpLayout), reshardedEdges(reshardedEdges) {}
+};
+
+class ShardSolver {
+private:
+  static constexpr size_t kNumBitsetBits = 128;
```

Comment:
Do you think we need 128?  We can probably get away with 64 which would fit word size and be significantly faster.

Reply:
Not sure, we can put anything for now and adjust. Since you had a bit stronger opinion than me, 64 it is.